File: server/lib/detect.js
'use strict';

function endsWithQuestion(text) {
  const t = String(text == null ? '' : text).trimEnd();
  return t.length > 0 && t[t.length - 1] === '?';
}

function isDigit(c) {
  return c >= '0' && c <= '9';
}

function parseChoices(text) {
  const out = [];
  for (const raw of String(text == null ? '' : text).split('\n')) {
    const line = raw.trim();
    let i = 0;
    while (i < line.length && isDigit(line[i])) i++;
    if (i === 0) continue;
    if (i >= line.length) continue;
    const sep = line[i];
    if (sep !== '.' && sep !== ')') continue;
    let j = i + 1;
    if (j >= line.length || line[j] !== ' ') continue;
    while (j < line.length && line[j] === ' ') j++;
    const label = line.slice(j).trim();
    if (!label) continue;
    out.push({ n: parseInt(line.slice(0, i), 10), label });
  }
  return out;
}

module.exports = { endsWithQuestion, parseChoices };

File: server-go/detect.go
package main

import (
	"strconv"
	"strings"
)

type Choice struct {
	N     int
	Label string
}

func endsWithQuestion(text string) bool {
	t := strings.TrimRight(text, " \n\t\r")
	return len(t) > 0 && t[len(t)-1] == '?'
}

func parseChoices(text string) []Choice {
	out := []Choice{}
	for _, raw := range strings.Split(text, "\n") {
		line := strings.TrimSpace(raw)
		i := 0
		for i < len(line) && line[i] >= '0' && line[i] <= '9' {
			i++
		}
		if i == 0 || i >= len(line) {
			continue
		}
		sep := line[i]
		if sep != '.' && sep != ')' {
			continue
		}
		j := i + 1
		if j >= len(line) || line[j] != ' ' {
			continue
		}
		for j < len(line) && line[j] == ' ' {
			j++
		}
		label := strings.TrimSpace(line[j:])
		if label == "" {
			continue
		}
		n, _ := strconv.Atoi(line[:i])
		out = append(out, Choice{N: n, Label: label})
	}
	return out
}
